File: units_cpp/src/unitsTest/cpp/constants/test_jupiter_properties.cc
/*
 * test_jupiter_properties.cc
 *
 *  Created on: Dec 27, 2017
 *      Author: sagis
 */

#include <units/constants/jupiter_properties.h>
#include <units/constants/physical_constants.h>
#include <units/units_multiply.h>
#include <units/units_ratio.h>
#include <units/units_pow.h>
#include <units/units_sqrt.h>
#include <units/planetary_constants_test_suite.h>

using namespace units::jupiter_properties;
using namespace units;
using units::physical_constants::gravitationalConstant;

class TestJupiterProperties : public PlanetaryConstantsTestSuite
{
protected:
	TestJupiterProperties()
	: PlanetaryConstantsTestSuite{6.5} // %
	{
	}
};

TEST_F(TestJupiterProperties, test_surface_gravity)
{
	auto calculatedSurfaceGravity =
			units_ratio<units::acceleration::MetersPerSquareSecond>(
					auto_units_multiply(gravitationalConstant, jupiterMass),
					auto_units_multiply(jupiterEquatorialRadius, jupiterEquatorialRadius));
	checkEqualValue(jupiterSurfaceGravity, calculatedSurfaceGravity);
}

TEST_F(TestJupiterProperties, test_mass)
{
	auto calculatedMass = (M_PI * 4 / 3) *
			units_multiply<units::mass::Kilograms>(
					jupiterMeanDensity ,units_pow<3>(jupiterEquatorialRadius));
	checkEqualValue(jupiterMass, calculatedMass);
}

TEST_F(TestJupiterProperties, test_escape_velocity)
{
	auto calculatedEscapeVelocity = units_sqrt<units::speed::MetersPerSecond> (
			2 * auto_units_ratio(
					auto_units_multiply(gravitationalConstant, jupiterMass),
					jupiterEquatorialRadius
				)
			);
	checkEqualValue(jupiterEscapeVelocity, calculatedEscapeVelocity);
}

File: units_java/src/test/java/com/units/tests/constants/TestJupiterProperties.java
package com.units.tests.constants;

import static com.units.constants.PhysicalConstants.gravitationalConstant;
import static com.units.constants.JupiterProperties.*;

import org.junit.Test;

import com.units.acceleration.MetersPerSquareSecond;
import com.units.mass.Kilograms;
import com.units.speed.MetersPerSecond;

public class TestJupiterProperties extends PlanetaryConstantsTestSuite {
	
	public TestJupiterProperties() {
		super(7.0); // %
	}
	
	@Test
	public void testSurfaceGravity() {
		MetersPerSquareSecond calculatedAcceleration =
				gravitationalConstant
				.multiply(jupiterMass)
				.divide(jupiterEquatorialRadius)
				.divide(jupiterEquatorialRadius)
				.castTo(MetersPerSquareSecond.class);
		checkAcceleration(jupiterSurfaceGravity, calculatedAcceleration);	
	}
	
	@Test
	public void testMass() {
		Kilograms calculatedMass =
				jupiterMeanDensity
				.multiply(jupiterEquatorialRadius.pow(3))
				.castTo(Kilograms.class)
				.multiplyByScalar(Math.PI * 4 / 3);
		checkMass(jupiterMass, calculatedMass);
	}
	
	@Test
	public void testEscapeVelocity() {
		MetersPerSecond calculatedEscapeVelocity =
				gravitationalConstant
				.multiply(jupiterMass)
				.divide(jupiterEquatorialRadius)
				.sqrt()
				.castTo(MetersPerSecond.class)
				.multiplyByScalar(Math.sqrt(2));
		checkVelocity(jupiterEscapeVelocity, calculatedEscapeVelocity);
	}

}
